### ISSUE TITLE
tflite: best_int8.tflite does not generate correct bounding box.

### ISSUE THREAD

**Issue by harufumigithub** [ultralytics/ultralytics#21030]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

yolo export model=yolo11n.pt format=tflite int8=true
generates followings: 
best_saved_model$ tree
.
├── assets
├── best_float16.tflite
├── best_float32.tflite
├── best_full_integer_quant.tflite
├── best_integer_quant.tflite
├── blob_detr_int8_v50.tflite
├── fingerprint.pb
├── metadata.yaml
├── saved_model.pb

Among these tflite, running inference works well except best_int8.tflite. 
best_int8.tflite place wrong bounding box.  When this happens, you need to reset or restart the environment and solve issue. However, this wrong BBox placement comes back time to time.  

### Environment

yolov8$ yolo check
Ultralytics 8.3.68 🚀 Python-3.9.13 torch-2.5.1+cu124 CUDA:0 (NVIDIA GeForce RTX 3070 Laptop GPU, 7747MiB)
Setup complete ✅ (16 CPUs, 62.4 GB RAM, 183.0/245.0 GB disk)

OS                  Linux-6.8.0-60-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.9.13
Install             pip
RAM                 62.45 GB
Disk                183.0/245.0 GB
CPU                 11th Gen Intel Core(TM) i7-11800H 2.30GHz
CPU count           16
GPU                 NVIDIA GeForce RTX 3070 Laptop GPU, 7747MiB
GPU count           1
CUDA                12.4

numpy               ✅ 1.23.5>=1.23.0
numpy               ✅ 1.23.5<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.5.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 9.2.0>=7.1.2
pyyaml              ✅ 6.0>=5.3.1
requests            ✅ 2.28.1>=2.23.0
scipy               ✅ 1.10.1>=1.4.1
torch               ✅ 2.5.1+cu124>=1.8.0
torch               ✅ 2.5.1+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.1+cu124>=0.9.0
tqdm                ✅ 4.64.1>=4.64.0
psutil              ✅ 5.9.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 1.4.4>=1.1.4
seaborn             ✅ 0.11.2>=0.11.0
ultralytics-thop    ✅ 2.0.10>=2.0.0

### Minimal Reproducible Example

yolo export model=yolo11n.pt format=tflite int8=true

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @harufumigithub, thank you for your detailed report and for using Ultralytics 🚀! We appreciate you taking the time to provide your environment details and the steps you've followed.

This is an automated response to help streamline troubleshooting—an Ultralytics engineer will also review and assist you soon.

If this is a 🐛 Bug Report, please ensure you've included a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/)—it looks like you have, which is great! If you can share any additional sample input images, code snippets, or logs that demonstrate the incorrect bounding box outputs with `best_int8.tflite`, that will further help us investigate.

You can find more information and usage examples in our [Docs](https://docs.ultralytics.com/), including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) guides.

Join the Ultralytics community for real-time chat on [Discord](https://discord.com/invite/ultralytics) 🎧, dive into in-depth discussions on [Discourse](https://community.ultralytics.com/), or check out our [Subreddit](https://reddit.com/r/Ultralytics) for more support and knowledge sharing.

## Upgrade

To help rule out issues already addressed in recent updates, please upgrade to the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for reporting this! An Ultralytics engineer will follow up with you soon.

**Comment by Y-T-G** [comment]
You also need to provide the inference code you're using to reproduce this. Are you using Ultralytics for inference? Or your own custom inference code?

**Comment by harufumigithub** [comment]
best.pt, best_integer_quant.pt and other works fine except best_int8.pt  Bounding box appears incorrectly. If you need a sample image and best.pt along with converted tflites I have experimented, I could sent you.  Please let me know.

conversion is done in the following code followed by the inference script:
yolo export model=runs/segment/train/weights/best.pt format=tflite int8=true 

model = YOLO(k_model_path)
results = model.predict(source = file, save=True, conf=conf, show_labels=True, show_conf=True, save_txt=True, device=0)

**Comment by glenn-jocher** [comment]
Thanks for the details! The issue with `best_int8.tflite` producing incorrect bounding boxes is likely related to the INT8 quantization calibration process. Since you're using the standard Ultralytics inference code, this suggests the issue occurs during export.

A few questions to help debug this:

1. Are you providing a `data` argument during export? INT8 quantization requires calibration data - if not specified, it defaults to a small dataset which might not be representative of your use case.

2. Can you try exporting with explicit calibration data: `yolo export model=runs/segment/train/weights/best.pt format=tflite int8=true data=path/to/your/dataset.yaml`

3. What specific incorrect behavior are you seeing with the bounding boxes? Are they completely misplaced, scaled incorrectly, or showing at wrong coordinates?

The intermittent nature suggests the calibration process might be inconsistent. You could also try setting a fixed random seed or using a larger calibration dataset with the `fraction` parameter to improve quantization stability.

**Comment by harufumigithub** [comment]
Hi glenn,

Thanks for the prompt feedback.
1. I did provided data argument.  I also tried without providing data argument.  Both does not work well.
2. I did provided data argument.  I also tried without providing data argument.  Both does not work well.
3. Incorrect bounding box location/coordinates, size, scaling.

I have see this working on my environment with best.pt but not converted one best_int8.tflite. It is interesting that best_integer_quant.tflite does not have this issue.  It is intermittent.  Our current solution for now is to use best_float16.tflie since  best_integer_quant loses accuracy quite a bit for small object detection.

**Comment by harufumigithub** [comment]
When best_int8.tflite is not working, it output large scale bounding box with confidence = 1.0  everywhere.  This happens both w/ and w/o data argument.yaml.  Some time the correct size but positions are completely off with confidence = 1.0.

**Comment by glenn-jocher** [comment]
Thanks for the additional details! The issue you're describing - large bounding boxes with confidence = 1.0 everywhere - is characteristic of a poorly calibrated INT8 model. The key insight is that `best_integer_quant.tflite` works but `best_int8.tflite` doesn't.

Looking at the export process, `best_int8.tflite` is actually the dynamic range quantized model (renamed from `*_dynamic_range_quant.tflite`), while `best_integer_quant.tflite` is the full integer quantized model. These use different quantization approaches.

Try these steps:

1. Delete any existing calibration cache files (`.cache` files in your model directory) before re-exporting
2. Since `best_integer_quant.tflite` works correctly, consider using that instead - it's also INT8 quantized but uses a different calibration method
3. If you need the dynamic range quantization specifically, try increasing the calibration dataset size with `fraction=1.0` to ensure full dataset usage

The intermittent nature suggests the calibration process for dynamic range quantization may be unstable. Using `best_integer_quant.tflite` might be the most reliable solution for your use case.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in the pull request ultralytics 8.3.202 TFLite per-channel INT8 quantization fix by @Y-T-G with contributions from @glenn-jocher. The change updates TFLite export behavior to better support INT8 per-channel quantization and improve GPU delegate reliability.

Key changes that may resolve the problem:
- Package bumped to 8.3.202.
- Removed the explicit per-tensor quantization setting so TFLite can apply per-channel INT8 where appropriate.
- Made BatchMatMul unfolding conditional: enabled for non-INT8 exports and disabled when using INT8, helping avoid INT8 activation issues and improving detection counts on GPU delegates.
- Addresses the instability discussed in the INT8-FP16 activation bug thread. See: INT8-FP16 activation bug discussion.

How to test the fix:
- Upgrade to the latest release with:
  - pip install -U ultralytics>=8.3.202
- Or install from the main branch to test the very latest changes:
  - pip install git+https://github.com/ultralytics/ultralytics.git@main
- Example export command for INT8 TFLite:
  - yolo export model=path/to/model.pt format=tflite int8=True

Feedback on whether this resolves the issue would be greatly appreciated. Thanks for reporting this, and please share any remaining problems or edge cases so they can be addressed.
